src/webgpu/shader/execution/shader_io/fragment_builtins.spec.ts

```diff
@@ -525,46 +556,64 @@ async function renderFragmentShaderInputsTo4TexturesAndReadbackValues(
   pass.end();
   t.queue.submit([encoder.finish()]);
 
-  const buffers = await Promise.all(
-    textures.map(async texture => {
-      const buffer = copyRGBA8EncodedFloatTextureToBufferIncludingMultisampledTextures(t, texture);
-      await buffer.mapAsync(GPUMapMode.READ);
-      return new Float32Array(buffer.getMappedRange());
-    })
-  );
-  const numElements = buffers[0].length;
-  const data = new Float32Array(numElements * 4);
-  for (let i = 0; i < numElements; ++i) {
-    const offset = i * 4;
-    data[offset + 0] = buffers[0][i];
-    data[offset + 1] = buffers[1][i];
-    data[offset + 2] = buffers[2][i];
-    data[offset + 3] = buffers[3][i];
-  }
-  return data;
+  const buffer = copyRGBA8EncodedFloatTexturesToBufferIncludingMultisampledTextures(t, textures);
+  await buffer.mapAsync(GPUMapMode.READ);
+  return new Float32Array(buffer.getMappedRange());
 }
 
-/**
- * Check whether two TypeArrays are approximately equal
- * Returns `undefined` if the check passes, or an `Error` if not.
- */
-function checkElementsApproximatelyEqual(
-  actual: TypedArrayBufferView,
-  expected: TypedArrayBufferView,
-  maxDiff = 0.000001
-) {
-  return checkElementsPassPredicate(
-    actual,
-    (index, value) => Math.abs(value - expected[index]) <= maxDiff,
-    {
-      predicatePrinter: [
-        {
-          leftHeader: 'expected ==',
-          getValueForCell: i => expected[i],
-        },
-      ],
-    }
+function checkSampleRectsApproximatelyEqual({
+  width,
+  height,
+  sampleCount,
+  actual,
+  expected,
+  maxFractionalDiff,
+}: {
+  width: number;
+  height: number;
+  sampleCount: number;
+  actual: Float32Array;
+  expected: Float32Array;
+  maxFractionalDiff: number;
+}) {
+  const subrectOrigin = [0, 0, 0];
+  const subrectSize = [width, height, 1];
+  const areaDesc = {
+    bytesPerRow: width * sampleCount * 4 * 4,
+    rowsPerImage: height,
+    subrectOrigin,
+    subrectSize,
+  };
+
+  const format = 'rgba32float';
+  const actTexelView = TexelView.fromTextureDataByReference(
+    format,
+    new Uint8Array(actual.buffer),
+    areaDesc
+  );
+  const expTexelView = TexelView.fromTextureDataByReference(
+    format,
+    new Uint8Array(expected.buffer),
+    areaDesc
+  );
+
+  const failedPixelsMessage = findFailedPixels(
+    format,
+    { x: 0, y: 0, z: 0 },
+    { width, height, depthOrArrayLayers: 1 },
+    { actTexelView, expTexelView },
+    { maxFractionalDiff }
```

Comment:
Hmm, I think that there is a preference to use ULPs if possible. Given the fractional diff you are using I think that it's likely within 1 ULP.